cmake: Use BUILD_INTERFACE with TORCH_SRC_DIR

seemethere opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#60403 cmake: Use BUILD_INTERFACE with TORCH_SRC_DIR**

TORCH_SRC_DIR has the potential to be hardcoded if used by itself so
prefer to use along with BUILD_INTERFACE so that the actual path is not
hardcoded

See https://cmake.org/cmake/help/latest/command/target_include_directories.html

Fixes errors initially introduced in https://github.com/pytorch/pytorch/pull/59563

Signed-off-by: Eli Uriegas <eliuriegas@fb.com>

Differential Revision: [D29276503](https://our.internmc.facebook.com/intern/diff/D29276503)

seemethere committed:
cmake: Prefer CMAKE_CURRENT_SOURCE_DIR over TORCH_SRC_DIR

TORCH_SRC_DIR has the potential to be hardcoded thus breaking downstream
cmake extensions. Prefer CMAKE_CURRENT_SOURCE_DIR with BUILD_INTERFACE
to make it magically work together

See https://cmake.org/cmake/help/latest/command/target_include_directories.html

Signed-off-by: Eli Uriegas <eliuriegas@fb.com>

[ghstack-poisoned]

driazati commented on caffe2/CMakeLists.txt:
```diff
@@ -1187,7 +1187,7 @@ endif()
 # jit/unpickler.cpp need to be compiled only when USE_DISTRIBUTED is set
 if(USE_DISTRIBUTED)
   # Needed to support the inclusion of c10d/Foo.hpp headers.
-  target_include_directories(torch_cpu PUBLIC ${TORCH_SRC_DIR}/lib)
```
is there ever a reason to use `TORCH_SRC_DIR` (cc @malfet)? maybe we should just get rid of it to prevent problems like this in the future

seemethere replied:
I think for other cmake files we basically use `TORCH_SRC_DIR` as an alias for `CMAKE_CURRENT_SOURCE_DIR` but maybe we shouldn't any more.

Also for some reason caffe2 has a different definition of what TORCH_SRC_DIR actually is:
https://github.com/pytorch/pytorch/blob/65f33ec85c2a7d8fb9bf582017d3170bf89e6c12/caffe2/CMakeLists.txt#L325-L330

Compared with the torch cmake:
https://github.com/pytorch/pytorch/blob/65f33ec85c2a7d8fb9bf582017d3170bf89e6c12/torch/CMakeLists.txt#L19-L22

seemethere replied:
Actually I think the mistake here was that the original author didn't include `BUILD_INTERFACE` which should automatically set it to not be hardcoded

seemethere commented:
@seemethere has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29276503).

seemethere committed:
Update on "cmake: Prefer CMAKE_CURRENT_SOURCE_DIR over TORCH_SRC_DIR"


TORCH_SRC_DIR has the potential to be hardcoded thus breaking downstream
cmake extensions. Prefer CMAKE_CURRENT_SOURCE_DIR with BUILD_INTERFACE
to make it magically work together

See https://cmake.org/cmake/help/latest/command/target_include_directories.html

Signed-off-by: Eli Uriegas <eliuriegas@fb.com>

Differential Revision: [D29276503](https://our.internmc.facebook.com/intern/diff/D29276503)

[ghstack-poisoned]

seemethere committed:
Update on "cmake: Use BUILD_INTERFACE with TORCH_SRC_DIR"


TORCH_SRC_DIR has the potential to be hardcoded thus breaking downstream
cmake extensions. Prefer CMAKE_CURRENT_SOURCE_DIR with BUILD_INTERFACE
to make it magically work together

See https://cmake.org/cmake/help/latest/command/target_include_directories.html

Signed-off-by: Eli Uriegas <eliuriegas@fb.com>

Differential Revision: [D29276503](https://our.internmc.facebook.com/intern/diff/D29276503)

[ghstack-poisoned]

seemethere commented:
@seemethere has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29276503).
